src/firebase-db.ts

```diff
@@ -494,6 +502,62 @@ export function createOrUpdateAnswer(answer: IExportableAnswerMetadata) {
   return firestoreSetPromise;
 }
 
+
+export const saveInteractiveStateHistoryEntry = async (options: ISaveInteractiveStateHistoryEntryOptions) => {
+  if (!portalData) {
+    return;
+  }
+  const { answerId, questionId, interactiveStateHistoryId, state } = options;
+
+  let historyEntry: IInteractiveStateHistory;
+  const baseHistoryEntry: IInteractiveStateHistoryBaseEntry = {
+    id: interactiveStateHistoryId,
+    answer_id: answerId,
+    question_id: questionId,
+    state_type: "full",
+    created_at: firebase.firestore.FieldValue.serverTimestamp()
+  };
+  if (portalData.type === "authenticated") {
+    const authenticatedEntry: IAuthenticatedInteractiveStateHistoryEntry = {
+      ...baseHistoryEntry,
+      type: "authenticated",
+      context_id: portalData.contextId,
+      platform_id: portalData.platformId,
+      platform_user_id: portalData.platformUserId.toString(),
+      resource_link_id: portalData.resourceLinkId,
+      run_key: "",
+    };
+    historyEntry = authenticatedEntry;
+  } else {
+    const anonymousEntry: IAnonymousInteractiveStateHistoryEntry = {
+      ...baseHistoryEntry,
+      type: "anonymous",
+      run_key: portalData.runKey,
+    };
+    historyEntry = anonymousEntry;
+  }
+
+  // actual state to be stored separately with a kind field to allow for different storage strategies
+  // in the future
+  const historyEntryWithState: IInteractiveStateHistoryWithState = {
+    ...historyEntry,
+    state,
+  };
+
+  const historyEntrySetPromise = app.firestore()
+    .doc(interactiveStateHistoryPath(interactiveStateHistoryId))
+    .set(historyEntry as Partial<firebase.firestore.DocumentData>);
+
+  const historyEntryStateSetPromise = app.firestore()
+    .doc(interactiveStateHistoryStatePath(interactiveStateHistoryId))
+    .set(historyEntryWithState as Partial<firebase.firestore.DocumentData>);
```

Comment:
It'd be safer to use a "batch" here: https://firebase.google.com/docs/firestore/manage-data/transactions#batched-writes
That way the metadata won't get written and then some error or disconnection causes it to be missing its state.